enhancements/network/on-prem-bgp-networking.md

```diff
@@ -0,0 +1,562 @@
+It can be a single IPv4 or IPv6 address, or comprised of both an IPv4 and an IPv6 address, in which case BGP must advertise both addresses.
+The external wildcard `*.apps.$cluster_name.$base-domain` DNS record should point to Ingress VIPs.
+
+
+### Highly available routed VIPs
+Access to the Kubernetes API (port 6443) and Ingress from clients both external
+and internal to the cluster should be highly available across control plane machines.
+
+The API VIPs first resides on the bootstrap instance and then move to all control plane nodes. The move will happen once the bootstrap process has completed, a FRR sidecar container residing on the boostrap will remove the VIP and FRR will update
+the leaf routing that it doesn't have the VIP anymore. 
+
+The Ingress VIPs will be created on the nodes where the Ingress Controller reports healthy.
+
+FRR will be deployed as a [static pod](https://kubernetes.io/docs/tasks/administer-cluster/static-pod/) and the assets will be rendered by the Machine Config Operator.
+When replacing a control plane machine, a new BGP session will be established with the peers. This causes no issues with the existing sessions by nature of BGP.
+
+#### Init containers
+
+- `copy-frr-static-configs`: This container will create two static files in `/etc/frr`, so FRR can be started and monitored:
+  - daemons: this file will contain the daemons that FRR will run.
+  - vtysh.conf: this file is required so the healthcheck can run (it will be used to check if FRR is running).
+- `verify-api-int-resolvable`: This container already exists for Keepalived/HAProxy and helps to ensure that the internal VIP can be resolved before creating it.
+- `render-config-frr`: This container will use baremetal-runtimecfg to render `frr.conf` based on a template that will be provided in MCO. Below we provide details on this file.
+
+#### FRR container
+
+This container will handle the start, stop and reload of BGP and BFD daemons in FRR.
+
+- it will wait for the `frr.conf` configuration file to exist (generated by baremetal-runtimecfg) and then start BGP and BFD daemons once it exists.
+- it will stop daemons if it gets a `stop` message via a socat handler running with baremetal-runtimecfg, like it's already done for HAProxy & Keepalived (and re-using the same conditions if we need to stop it for
+  feature parity). This part is already implemented for CoreDNS, HAProxy and Keepalived.
+- it will reload daemons if it gets a `reload` message from the socat handler, if configuration on-disk was changed with the one currently running.
+
+FRR's configuration will be opinionated but flexible enough to meet customer needs. The BGP configuration will be defined like this:
+- A list of BGP peers (usually their Top-of-Rack routers) IPs
+- Each peer would have an IP address, a BGP password and an ASN
+
+The customers will not be able to customize FRR until we find some real limitations. BGP and BFD daemons will be enabled by `copy-frr-static-configs` container.
```

Comment:
Generally speaking we don't support much in the way of customization of the internal loadbalancer. If customers want more advanced loadbalancer configuration they should use an external one that they have full control over. Because of the need for this to run before the cluster is up it's a little awkward to make changes to it after deployment (new machine-configs would require a reboot of all affected nodes) so we intentionally avoid exposing that ability.

Reply:
I wish it existed but it doesn't, the doc is clear: https://docs.frrouting.org/en/latest/setup.html#reloading
FRR has to be reloaded when a new config needs to be applied.

Reply:
Good points. I'll add some words here.